crates/ruff_linter/src/rules/pylint/rules/redefined_argument_from_local.rs

```diff
@@ -0,0 +1,167 @@
+    fn message(&self) -> String {
+        let RedefinedArgumentFromLocal { name } = self;
+        format!("Redefining argument with the local name `{name}`")
+    }
+}
+
+#[derive(Default)]
+struct StoredNamesVisitor<'a> {
+    identifiers: Vec<&'a Identifier>,
+    expressions: Vec<&'a ast::ExprName>,
+}
+
+impl<'a> StoredNamesVisitor<'a> {
+    fn names(&self) -> Vec<(&str, TextRange)> {
+        self.identifiers
+            .iter()
+            .map(|i| (i.as_str(), i.range()))
+            .chain(self.expressions.iter().map(|e| (e.id.as_str(), e.range)))
+            .collect::<Vec<_>>()
+    }
+}
+
+/// `Visitor` to collect all stored names in a statement.
+impl<'a> Visitor<'a> for StoredNamesVisitor<'a> {
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        match stmt {
+            Stmt::For(ast::StmtFor { target, .. }) => self.visit_expr(target),
+            Stmt::Try(ast::StmtTry { handlers, .. }) => {
+                for handler in handlers {
+                    let ExceptHandler(ExceptHandlerExceptHandler { name, .. }) = handler;
+                    if let Some(ident) = name {
+                        self.identifiers.push(ident);
+                    }
+                }
+            }
+            Stmt::With(ast::StmtWith { items, .. }) => {
+                for item in items {
+                    if let Some(expr) = &item.optional_vars {
+                        self.visit_expr(expr);
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        match expr {
+            Expr::Name(name) => {
+                if name.ctx == ExprContext::Store {
+                    self.expressions.push(name);
+                }
+            }
+            _ => visitor::walk_expr(self, expr),
+        }
+    }
+}
+
+/// PLR1704
+pub(crate) fn redefined_argument_from_local(checker: &mut Checker, stmt: &Stmt) {
+    let mut visitor = StoredNamesVisitor::default();
+    visitor.visit_stmt(stmt);
+
+    let mut dianostics = vec![];
+    let mut already_added: FxHashSet<TextRange> = FxHashSet::default();
+    let mut scope = checker.semantic().current_scope();
+    loop {
+        if let ScopeKind::Function(ast::StmtFunctionDef { parameters, .. }) = scope.kind {
+            for (name, range) in visitor.names() {
+                if already_added.contains(&range) {
+                    continue;
+                }
+                if parameters.includes(name) {
+                    dianostics.push(Diagnostic::new(
+                        RedefinedArgumentFromLocal {
+                            name: name.to_string(),
+                        },
+                        range,
+                    ));
+                    already_added.insert(range);
+                }
+            }
+        }
+        if let Some(scope_id) = scope.parent {
+            scope = &checker.semantic().scopes[scope_id];
+        } else {
+            break;
+        };
```

Comment:
I'm wondering if there's a slightly different way to structure this rule. Bear with me...

Instead of analyzing the AST, I think we can achieve the same effect by looking at _bindings_.

When we have something like:

```python
def f(x):
    for x in y:
        ...
```

We create a binding for the `x` in `def f(x)`, and then later, a binding for `x` in `for x in y`, and we track that this second binding shadows the first.

If you look in `deferred_scopes.rs`, you'll see we have some rules based on looking at these shadows:

```rust
if checker.enabled(Rule::ImportShadowedByLoopVar) {
    for (name, binding_id) in scope.bindings() {
        for shadow in checker.semantic.shadowed_bindings(scope_id, binding_id) {
            // If the shadowing binding isn't a loop variable, abort.
            let binding = &checker.semantic.bindings[shadow.binding_id()];
            if !binding.kind.is_loop_var() {
                continue;
            }

            // If the shadowed binding isn't an import, abort.
            let shadowed = &checker.semantic.bindings[shadow.shadowed_id()];
            if !matches!(
                shadowed.kind,
                BindingKind::Import(..)
                    | BindingKind::FromImport(..)
                    | BindingKind::SubmoduleImport(..)
                    | BindingKind::FutureImport
            ) {
                continue;
            }

            // If the bindings are in different forks, abort.
            if shadowed.source.map_or(true, |left| {
                binding.source.map_or(true, |right| {
                    checker.semantic.different_branches(left, right)
                })
            }) {
                continue;
            }

            #[allow(deprecated)]
            let line = checker.locator.compute_line_index(shadowed.start());

            checker.diagnostics.push(Diagnostic::new(
                pyflakes::rules::ImportShadowedByLoopVar {
                    name: name.to_string(),
                    line,
                },
                binding.range(),
            ));
        }
    }
}
```

I think this rule could be structured similarly: iterate over the pairs of shadowed bindings, and look for pairs in which the original binding is `BindingKind::Argument`, and the new binding is `BindingKind::LoopVar`.